### ISSUE TITLE
NGINX Configuration

### ISSUE THREAD

**Issue by yankeeinlondon** [Luracast/Restler#352]
When installing **Restler** under NGINX / FPM it is a very common scenario to have multiple "services"  or tech stacks being served on a single domain. For instance, you may have Restler providing the Restful API to an application while also hosting the main website for the application on the same URL. Let's assume that in this case the main application uses NODEJS and serves content off the root directory (`http://localhost`) and Restler is configured to work off the api directory (`http://localhost/api`). 

This might translate into a `nginx.conf` file similar to the following:

```
upstream application_upstream {
    server 127.0.0.1:2380;
    keepalive 64;
}

server {
    listen 1080;
    server_name localhost;
    access_log /var/log/nginx/access.log;
    root /path/to/website;

    # Drop requests to bad file types
    location ~ \.(aspx|jsp|cgi)$ {
        return 410;
    }

    # serves the web application via a TCP socket connection which is served by
    # a separate tech stack (in this case NodeJS)
    location / {
        proxy_redirect off;
        proxy_set_header    X-Real-IP           $remote_addr;
        proxy_set_header    X-Forwarded-For     $proxy_add_x_forwarded_for;
        proxy_set_header    X-Forwarded-Proto   $scheme;
        proxy_set_header    Host                $http_host:$server_port;
        proxy_set_header    X-NginX-Proxy       true;
        proxy_set_header    Connection          "";
        proxy_http_version  1.1;
        proxy_pass          http://application_upstream;
    }
    # serves the API Explorer as static files       
    location /explorer {
        index index.html;
    }

    # serves the Restler API via an FPM handoff
    location /api {
        gzip    off;
        fastcgi_pass    unix:/path/to/socket/api.socket;
        fastcgi_index   index.php;
        fastcgi_param   SCRIPT_FILENAME  $document_root$fastcgi_script_name;
        include         fastcgi_params;
    }
}       
```

Clearly the important part of this config file is the `location /api` and this mimics the example given on the Luracast website but with a few changes:
1. use of `$document_root` instead of hard-coded path (this is strongly recommended change)
2. elimination of the '/' character between the document root and the `$fastcgi_script_name` as the document root will already terminate with this character and this thereby eliminates the duplication.
3. this config uses Unix sockets instead of TCP sockets as this is far more efficient if you're on Unix/Linux/OSX
4. uses the `/api` pattern to match for FPM hand-off instead of `~ \.*php$` ... this is to isolate the PHP to only those URLs associated with the API. 

> note: on this last point, I have tried the suggeste `~ \.*php$` matching pattern too and that results in similar errors so this should not be seen as the problem area

The basic FPM gateway does work but there are several problems which I'm struggling with:

**Base Call | Use Case**
Calls to `http://localhost/api` or `http://localhost/api/index.php` both result in the successful execution of the index.php script. However, they also both result in a **404** error being thrown by the script:

```
{
    "error": {
        "code": 404,
        "message": "Not Found"
    },
    "debug": {
        "source": "Routes.php:436 at route stage",
        "stages": {
            "success": [
                "get"
            ],
            "failure": [
                "route",
                "negotiate",
                "message"
            ]
        }
    }
}
```

In order to better understand the "state" that is being passed into Restler's `index.php` file I added a debug statement that prints out the $_SERVER variable and this comes back as:

```
(
    [USER] => Ken
    [HOME] => /Users/Ken
    [FCGI_ROLE] => RESPONDER
    [SCRIPT_FILENAME] => /path/to/website/api/index.php
    [QUERY_STRING] => 
    [REQUEST_METHOD] => GET
    [CONTENT_TYPE] => 
    [CONTENT_LENGTH] => 
    [SCRIPT_NAME] => /api/index.php
    [REQUEST_URI] => /api/index.php
    [DOCUMENT_URI] => /api/index.php
    [DOCUMENT_ROOT] => /path/to/website
    [SERVER_PROTOCOL] => HTTP/1.1
    [GATEWAY_INTERFACE] => CGI/1.1
    [SERVER_SOFTWARE] => nginx/1.6.0
    [REMOTE_ADDR] => 127.0.0.1
    [REMOTE_PORT] => 56618
    [SERVER_ADDR] => 127.0.0.1
    [SERVER_PORT] => 1080
    [SERVER_NAME] => localhost
    [REDIRECT_STATUS] => 200
    [HTTP_HOST] => localhost:1080
    [HTTP_CONNECTION] => keep-alive
    [HTTP_PRAGMA] => no-cache
    [HTTP_CACHE_CONTROL] => no-cache
    [HTTP_ACCEPT] => text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
    [HTTP_USER_AGENT] => Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/37.0.2062.94 Safari/537.36
    [HTTP_ACCEPT_ENCODING] => gzip,deflate,sdch
    [HTTP_ACCEPT_LANGUAGE] => en-US,en;q=0.8
    [HTTP_COOKIE] => _ga=GA1.2.65956760.1408091744; PHPSESSID=q8p4hbsl859er7ftjg8ijisk97
    [PHP_SELF] => /api/index.php
    [REQUEST_TIME_FLOAT] => 1410345409.3903
    [REQUEST_TIME] => 1410345409
)
```

**Specific Resources | Use Case**
Things get worse when we try to get a specific resource like `resources.json` (for the API Explorer). In this example there are actually two sub-cases:
1. If we call `http://localhost/api/resources.json` we get a FastCGI error which indicates that the `index.php` is not being picked up:
   
   > [error] 83196#0: *545 FastCGI sent in stderr: "Primary script unknown" while reading response header from upstream
2. If we call `http://localhost/api/index.php/resources.json` we get the same **404** error as we got in the base case but obviously the $_SERVER variables are picking up the variation in the URL request

**IN SUMMARY**

While some of these issues are really NGINX or FPM issues, I think it's really important to have a good "out of the box" configuration for Restler developers to refer to and hence why I'm putting this in as a "issue" here. I am not very experienced in NGINX config so maybe I'm making some obvious mistakes but personally I would love some help in sorting this out.
